### PR TITLE
Rule Manager Floodlight integration

### PR DESCRIPTION
**Pull request by timofei-durakov**

### CODE REVIEW

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        List<GroupSpeakerData> commandData = new ArrayList<>();
+        for (OFGroupDescStatsEntry entry : statsReply.getEntries()) {
+        }
+        return commandData;
+    }
+
+    /**
+     * Convert group command data into OfGroupMod representation.
+     */
+    public OFGroupAdd convertInstallGroupCommand(GroupSpeakerData commandData, OFFactory ofFactory) {
```
Comment (rozdy) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java:94]: Move group related methods to separate converter or rename current one

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+     */
+    public OFGroupDelete convertDeleteGroupCommand(GroupSpeakerData commandData, OFFactory ofFactory) {
+        return ofFactory.buildGroupDelete()
+                .setGroup(OFGroup.of((int) commandData.getGroupId().getValue()))
+                .setGroupType(toOfGroupType(commandData.getType()))
+                .build();
+    }
+
+    private OFBucket toOfBucket(OFFactory ofFactory, Bucket bucket) {
+        List<OFAction> ofActions = OfInstructionsConverter.INSTANCE.convertApplyActions(bucket.getWriteActions(),
```
Comment (rozdy) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java:116]: Rename `convertApplyActions` -> `convertActions`. This method may be used for applyActions and writeActions.

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        }
+        return builder.build();
+    }
+
+    private OFPort toOfWatchPort(WatchPort watchPort) {
+        switch (watchPort) {
+            case ANY:
+                return OFPort.ANY;
+            default:
+                throw new IllegalArgumentException("Unknown watch port");
```
Comment (rozdy) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java:132]: Add details to error message. `String.format("Unknown watch port %s", watchPort)`

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        }
+    }
+
+    private OFPort fromOfPort(OFPort watchPort) {
```
Comment (rozdy) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java:136]: Probably method result type should be WatchPort

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        if (watchPort.equals(OFGroup.ANY)) {
+            return OFPort.ANY;
+        } else {
+            throw new IllegalArgumentException("Unknown watch group");
```
Comment (rozdy) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfFlowModConverter.java:140]: Add details to all error messages in this converter

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfMatchConverter.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                    .build());
+        } else {
+            OFMetadata metadata = match.get(MatchField.METADATA);
+            if (metadata != null) {
+                fieldMatches.add(FieldMatch.builder()
+                        .field(Field.METADATA)
+                        .value(metadata.getValue().getValue())
+                        .build());
+            }
+        }
```
Comment (rozdy) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/converter/rulemanager/OfMatchConverter.java:157]: IN_PORT match is missing here

**Review by dpoltavets — APPROVED** [review]

File: src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfSpeakerService.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Override
+    public void processBatchDelete(DeleteSpeakerCommandsRequest request, String key) {
+        SwitchId switchId = request.getSwitchId();
+        DatapathId dpId = DatapathId.of(switchId.toLong());
+        IOFSwitch sw = iofSwitchService.getSwitch(dpId);
+        OfBatchHolder holder = new OfBatchHolder(iofSwitchService, request.getMessageContext());
+        for (OfCommand data : request.getCommands()) {
+            data.buildInstall(holder, switchId);
```
Comment (dpoltavets) [inline src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfSpeakerService.java:78]: buildDelete

**Review by rozdy — APPROVED** [review]

**Review by surabujin — CHANGES_REQUESTED** [review]
The strongest objection from me is incorrect "recussive" in flows installation stages. It must be correct, or must be crystall clean described in extremmely notable place how "special" it implemented with all limitation. (Correct implementation ~10-15 LOC is much simplier that such description).

And the second objection - analyse of OF match object, it definetly must be reworked.

**Review by surabujin — APPROVED** [review]

**Pull request merged**